fractal_tasks_core/lib_ngff.py

```diff
@@ -0,0 +1,412 @@
+"""
+Pydantic models related to OME-NGFF 0.4 specs.
+"""
+import logging
+from typing import Literal
+from typing import Optional
+from typing import Union
+
+import zarr
+from pydantic import BaseModel
+from pydantic import Field
+from pydantic import validator
+
+
+class Window(BaseModel):
+    """
+    Model for `Channel.window`.
+
+    Note that we deviate by NGFF specs by making `start` and `end` optional.
+    See https://ngff.openmicroscopy.org/0.4/#omero-md.
+    """
+
+    max: float
+    min: float
+    start: Optional[float] = None
+    end: Optional[float] = None
+
+
+class Channel(BaseModel):
+    """
+    Model for an element of `Omero.channels`.
+
+    See https://ngff.openmicroscopy.org/0.4/#omero-md.
+    """
+
+    window: Window
+    label: Optional[str] = None
+    family: Optional[str] = None
+    color: str
+    active: Optional[bool] = None
+
+
+class Omero(BaseModel):
+    """
+    Model for `NgffImageMeta.omero`.
+
+    See https://ngff.openmicroscopy.org/0.4/#omero-md.
+    """
+
+    channels: list[Channel]
+
+
+class Axis(BaseModel):
```

Comment:
Good catch, thanks!

The relevant portion of the specs schema (https://github.com/ome/ngff/blob/main/0.4/schemas/image.schema#L111-L166) includes both the `items` and `contains` keywords.

After debugging a bit further, it turns out that https://koxudaxi.github.io/datamodel-code-generator/ is not handling `contains` in the way I would expect - see minimal working example here:
```
$ cat simpleschema.json 
{
   "type": "array",
   "contains": {
     "type": "number"
   }
}

$ datamodel-codegen --input simpleschema.json --output simpleschema.py --input-file-type jsonschema --output-model-type pydantic.BaseModel --use-standard-collections --target-python-version 3.10

$ cat simpleschema.py 
# generated by datamodel-codegen:
#   filename:  simpleschema.json
#   timestamp: 2023-09-27T07:33:44+00:00

from __future__ import annotations
from pydantic import BaseModel

class Model(BaseModel):
    __root__: list
```

This is trivial to fix in our PR, but we should be aware of the issue if we ever use the datamodel-generator tool in the future.
I just grepped for other `contains` in the relevant NGFF schemas (i.e. the ones I used in this PR), and there is only another one (it is in `coordinateTransormations`, but it requires no further action because it was a "placeholder" model which I simply removed).

Reply:
Fixed with 716cb29130de0255f0fe81ff160e26d97ab8c72b